generator/interfaces/gen.go

```diff
@@ -0,0 +1,167 @@
+package interfaces
+
+import (
+	"bytes"
+	"go/ast"
+	"go/printer"
+	"strings"
+
+	"github.com/devfile/api/generator/genutils"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"github.com/elliotchance/orderedmap"
+)
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/helpgen generate:headerFile=../header.go.txt,year=2020 paths=.
+
+var (
+	toplevelListMarker = markers.Must(markers.MakeDefinition("devfile:toplevellist", markers.DescribesField, struct{}{}))
+)
+
+type copyStructs struct {
+	StructName string
+	Fields     []string
+}
+
+// +controllertools:marker:generateHelp
+
+// Generator generates GO source code required for the API
+//
+// Generated source code mainly consists in interface implementations to manage unions and keyed top-level lists
+type Generator struct{}
+
+// RegisterMarkers registers the markers of the Generator
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	if err := markers.RegisterAll(into, toplevelListMarker); err != nil {
+		return err
+	}
+	into.AddHelp(toplevelListMarker,
+		markers.SimpleHelp("Devfile", "indicates that a given field of the Devfile body structure is a top-level list that should be managed through strategic merge patch during parent of plugin overriding."))
+	return genutils.RegisterUnionMarkers(into)
+}
+
+// Generate generates the artifacts
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	for _, root := range ctx.Roots {
+		ctx.Checker.Check(root, func(node ast.Node) bool {
+			// ignore interfaces
+			_, isIface := node.(*ast.InterfaceType)
+			return !isIface
+		})
+
+		root.NeedTypesInfo()
+
+		unions := orderedmap.NewOrderedMap()
+		toplevelListContainers := orderedmap.NewOrderedMap()
+		keyed := orderedmap.NewOrderedMap()
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			if info.Markers.Get(genutils.UnionMarker.Name) != nil {
+				unions.Set(info.Name, info)
+				return
+			}
+			for i, field := range info.Fields {
+				if field.Markers.Get(toplevelListMarker.Name) != nil {
+					toplevelListContainers.Set(info.Name, info)
+					if arrayType, isArrayType := field.RawField.Type.(*ast.ArrayType); isArrayType {
+						if ident, isIdent := arrayType.Elt.(*ast.Ident); isIdent {
+							keyed.Set(ident.Name, &info.Fields[i])
+						}
+					}
+				}
+			}
+		}); err != nil {
+			root.AddError(err)
+			return nil
+		}
+
+		genutils.WriteFormattedSourceFile("keyed_definitions", ctx, root, func(buf *bytes.Buffer) {
```

Comment:
Unhandled error  is returned here and seems there is no usage of that func with handled error.

Reply:
Fixed in commit https://github.com/devfile/api/pull/133/commits/7a6fc780cf9a34af3a95f1184aca4fd6b2d8fd43
I simply remove the returned value in the function, since it already registers errors inside the current root.